synapse/push/httppusher.py

```diff
@@ -13,60 +13,219 @@
+        # off as None though as we don't know any better.
+        self.max_stream_ordering = None
+
+        if 'data' not in pusherdict:
+            raise PusherConfigException(
+                "No 'data' key for HTTP pusher"
+            )
+        self.data = pusherdict['data']
+
+        self.name = "%s/%s/%s" % (
+            pusherdict['user_name'],
+            pusherdict['app_id'],
+            pusherdict['pushkey'],
         )
-        if 'url' not in data:
+
+        if 'url' not in self.data:
             raise PusherConfigException(
                 "'url' required in data for HTTP pusher"
             )
-        self.url = data['url']
-        self.http_client = _hs.get_simple_http_client()
+        self.url = self.data['url']
+        self.http_client = hs.get_simple_http_client()
         self.data_minus_url = {}
         self.data_minus_url.update(self.data)
         del self.data_minus_url['url']
 
     @defer.inlineCallbacks
-    def _build_notification_dict(self, event, tweaks, badge):
-        # we probably do not want to push for every presence update
-        # (we may want to be able to set up notifications when specific
-        # people sign in, but we'd want to only deliver the pertinent ones)
-        # Actually, presence events will not get this far now because we
-        # need to filter them out in the main Pusher code.
-        if 'event_id' not in event:
-            defer.returnValue(None)
+    def on_started(self):
+        yield self._process()
+
+    @defer.inlineCallbacks
+    def on_new_notifications(self, min_stream_ordering, max_stream_ordering):
+        with Measure(self.clock, "push.on_new_notifications"):
+            self.max_stream_ordering = max_stream_ordering
+            yield self._process()
+
+    @defer.inlineCallbacks
+    def on_new_receipts(self, min_stream_id, max_stream_id):
+        # Note that the min here shouldn't be relied upon to be accurate.
+
+        # We could check the receipts are actually m.read receipts here,
+        # but currently that's the only type of receipt anyway...
+        with Measure(self.clock, "push.on_new_receipts"):
+            badge = yield push_tools.get_badge_count(
+                self.hs.get_datastore(), self.user_id
+            )
+            yield self.send_badge(badge)
+
+    @defer.inlineCallbacks
+    def on_timer(self):
+        with Measure(self.clock, "push.on_timer"):
+            yield self._process()
 
-        ctx = yield self.get_context_for_event(event)
+    def on_stop(self):
+        if self.timed_call:
+            self.timed_call.cancel()
+
+    @defer.inlineCallbacks
+    def _process(self):
+        try:
+            self.processing = True
+            yield self._unsafe_process()
+        finally:
+            self.processing = False
+
+    @defer.inlineCallbacks
+    def _unsafe_process(self):
+        unprocessed = yield self.store.get_unread_push_actions_for_user_in_range(
```

Comment:
Can you document what `_unsafe_process` does? Or at least what is unsafe about, what it assumes?